eventqueue/processor.go

```diff
@@ -0,0 +1,229 @@
+	resetCh            chan struct{}
+	stopped            atomic.Bool
+}
+
+// NewProcessor returns a new Processor object.
+// executeFn is the callback invoked when the item is to be executed; this will be invoked in a background goroutine.
+func NewProcessor[T queueable](executeFn func(r T), clock kclock.Clock) *Processor[T] {
+	return &Processor[T]{
+		executeFn:          executeFn,
+		queue:              newQueue[T](),
+		processorRunningCh: make(chan struct{}, 1),
+		stopCh:             make(chan struct{}),
+		resetCh:            make(chan struct{}, 1),
+		clock:              clock,
+	}
+}
+
+// Enqueue adds a new item to the queue.
+// If a item with the same ID already exists, it'll be replaced.
+func (p *Processor[T]) Enqueue(r T) error {
+	if p.stopped.Load() {
+		return ErrProcessorStopped
+	}
+
+	// Insert or replace the item in the queue
+	// If the item added or replaced is the first one in the queue, we need to know that
+	p.queueLock.Lock()
+	peek, ok := p.queue.Peek()
+	isFirst := (ok && peek.Key() == r.Key()) // This is going to be true if the item being replaced is the first one in the queue
+	p.queue.Insert(r, true)
+	peek, _ = p.queue.Peek()         // No need to check for "ok" here because we know this will return an item
+	isFirst = isFirst || (peek == r) // This is also going to be true if the item just added landed at the front of the queue
+	p.process(isFirst)
+	p.queueLock.Unlock()
+
+	return nil
+}
+
+// Dequeue removes a item from the queue.
+func (p *Processor[T]) Dequeue(key string) error {
+	if p.stopped.Load() {
+		return ErrProcessorStopped
+	}
+
+	// We need to check if this is the next item in the queue, as that requires stopping the processor
+	p.queueLock.Lock()
+	peek, ok := p.queue.Peek()
+	p.queue.Remove(key)
+	if ok && peek.Key() == key {
+		// If the item was the first one in the queue, restart the processor
+		p.process(true)
+	}
+	p.queueLock.Unlock()
+
+	return nil
+}
+
+// Stop the processor.
+func (p *Processor[T]) Close() error {
+	if !p.stopped.CompareAndSwap(false, true) {
+		// Already stopped
+		return nil
+	}
+
+	// Send a signal to stop
+	close(p.stopCh)
+
+	// Blocks until process loop ends
+	t := time.NewTimer(5 * time.Second)
+	select {
+	case p.processorRunningCh <- struct{}{}:
+		// All good
+		if !t.Stop() {
+			<-t.C
+		}
+	case <-t.C:
+		return errors.New("Processor loop did not stop in 5s")
+	}
+
+	return nil
+}
+
+// Start the processing loop if it's not already running.
+// This must be invoked while the caller has a lock.
+func (p *Processor[T]) process(isNext bool) {
+	// Do not start a loop if it's already running
+	select {
+	case p.processorRunningCh <- struct{}{}:
+		// Nop - fallthrough
+	default:
+		// Already running
+		if isNext {
+			// If this is the next item, send a reset signal
+			// Use a select in case another goroutine is sending a reset signal too
+			select {
+			case p.resetCh <- struct{}{}:
+			default:
+			}
+		}
+		return
+	}
+
+	go p.processLoop()
```

Comment:
This should have a wait group to close can wait on this routine to return.

Reply:
Changed. Although we don't need a wg because we have `processorRunningCh`